File: src/main/scala/g0101_0200/s0101_symmetric_tree/Solution.scala
package g0101_0200.s0101_symmetric_tree

// #Easy #Top_100_Liked_Questions #Top_Interview_Questions #Depth_First_Search #Breadth_First_Search
// #Tree #Binary_Tree #Data_Structure_I_Day_11_Tree #Level_2_Day_15_Tree
// #Big_O_Time_O(N)_Space_O(log(N)) #2023_11_03_Time_454_ms_(90.38%)_Space_57.2_MB_(28.85%)

import com_github_leetcode.TreeNode

/*
 * Definition for a binary tree node.
 * class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {
 *   var value: Int = _value
 *   var left: TreeNode = _left
 *   var right: TreeNode = _right
 * }
 */
object Solution {
    def isSymmetric(root: TreeNode): Boolean = {
        if (root == null) {
            true
        } else {
            helper(root.left, root.right)
        }
    }

    private def helper(leftNode: TreeNode, rightNode: TreeNode): Boolean = {
        if (leftNode == null || rightNode == null) {
            leftNode == null && rightNode == null
        } else if (leftNode.value != rightNode.value) {
            false
        } else {
            helper(leftNode.left, rightNode.right) && helper(leftNode.right, rightNode.left)
        }
    }
}

File: src/main/python/g0101_0200/s0101_symmetric_tree/Solution.py
# #Easy #Top_100_Liked_Questions #Top_Interview_Questions #Depth_First_Search #Breadth_First_Search
# #Tree #Binary_Tree #Data_Structure_I_Day_11_Tree #Level_2_Day_15_Tree
# #Big_O_Time_O(N)_Space_O(log(N)) #2024_06_09_Time_35_ms_(71.46%)_Space_16.6_MB_(14.05%)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        if root is None:
            return True
        return self.helper(root.left, root.right)

    def helper(self, leftNode: TreeNode, rightNode: TreeNode) -> bool:
        if leftNode is None or rightNode is None:
            return leftNode is None and rightNode is None
        if leftNode.val != rightNode.val:
            return False
        return self.helper(leftNode.left, rightNode.right) and self.helper(leftNode.right, rightNode.left)
